testsuite/features/support/system_monitoring.rb

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) 2024 SUSE LLC.
+# Licensed under the terms of the MIT license.
+
+require 'time'
+
+# This method should return the last bootstrap duration for the given host
+# @param host [String] the hostname
+# @return [Float] the duration in seconds or nil if the duration could not be determined
+def last_bootstrap_duration(host)
+  system_name = get_system_name(host)
+  duration = nil
+  lines, _code = get_target('server').run('tail -n100 /var/log/rhn/rhn_web_api.log')
+
+  lines.each_line do |line|
+    if line.include?(system_name) && line.include?('systems.bootstrap')
+      match = line.match(/TIME: (\d+\.\d+) seconds/)
+      duration = match[1].to_f if match
+    end
+  end
+
+  raise ScriptError, "Boostrap duration not found for #{host}" if duration.nil?
+
+  duration
+end
+
+# This method should return the last onboarding duration for the given host
+# @param host [String] the hostname
+# @return [Float] the duration in seconds or nil if the duration could not be determined
+def last_onboarding_duration(host)
+  node = get_target(host)
+  system_id = get_system_id(node)
+  events = $api_test.system.get_event_history(system_id, 0, 10)
+  onboarding_events = events.select { |event| event['summary'].include? 'certs, channels, packages' }
+  last_event_id = onboarding_events.last['id']
+  event_details = $api_test.system.get_event_details(system_id, last_event_id)
+  Time.parse(event_details['completed']) - Time.parse(event_details['picked_up'])
+end
+
+# This method should return the synchronization duration for the given product
+# @param os_product_version [String] the product name
+# @return [Float] the duration in seconds or nil if the duration could not be determined
+def synchronization_duration(os_product_version)
+  channels_to_wait = CHANNEL_TO_SYNC_BY_OS_PRODUCT_VERSION.dig(product, os_product_version)
+  channels_to_wait = filter_channels(channels_to_wait, ['beta']) unless $beta_enabled
+  raise ScriptError, "Synchronization error, channels for #{os_product_version} in #{product} not found" if channels_to_wait.nil?
+
+  duration = 0
+  channel_to_evaluate = false
+  get_target('server').extract('/var/log/rhn/reposync.log', '/tmp/reposync.log')
+  File.foreach('/tmp/reposync.log') do |line|
+    if line.include?('Channel: ')
+      channel_name = line.split('Channel: ')[1].strip
+      channel_to_evaluate = channels_to_wait.include?(channel_name)
+    end
+    if line.include?('Total time: ') && channel_to_evaluate
+      match = line.match(/Total time: (\d+):(\d+):(\d+)/)
+      hours, minutes, seconds = match.captures.map(&:to_i)
+      total_seconds = (hours * 3600) + (minutes * 60) + seconds
+      duration += total_seconds
+      channel_to_evaluate = false
+    end
+  end
+  duration
+end
```

Comment:
That's risky if the file structure changes or something goes wrong with the file content. I don’t recommend adding file checks, as that could quickly require dozens lines of code. Instead, have you considered alternative options, like tracking the total process duration? That might be more reliable.

Reply:
There is not a unique process involved, as there are many channels synchronized.
The order how those channels are processes is not deterministic.
Tracking this it would require a different thread in our test framework or block the current thread processing it.
Additionally, that would require embedding that tracking in the middle of our tests from the moment we trigger the synchronization until it ends, while that way, it's much more independent, and it only need to process a file once we know all is synchronized.

Reply:
I am a bit worried what in case of timeout interruptions where file got cut, or packages woudn't be found and the time would be 0, and other corner cases like that. Can we at least use logging if there won't be regex match or it would be suspiciously different?

Reply:
I'm gonna raise an error if the duration is equal to zero, as that should never be the case.